sponsorship_compassion/models/account_invoice.py

```diff
@@ -124,65 +124,53 @@ def _group_or_split_reconcile(self):
         """
         partner = self.mapped('partner_id')
         partner.ensure_one()
-        total_amount = sum(self.mapped('amount_total'))
+        reconcile_amount = sum(self.mapped('amount_total'))
         move_lines = self.mapped('move_id.line_ids').filtered('debit')
         payment_search = [
             ('partner_id', '=', partner.id),
             ('account_id.code', '=', '1050'),
             ('reconciled', '=', False),
-            ('credit', '>', 0),
-            ('credit', '>', total_amount)
+            ('credit', '>', 0)
         ]
 
-        # First try to search a payment greater than the reconcile amount
         line_obj = self.env['account.move.line']
-        open_payment = line_obj.search(
-            payment_search, order='date asc', limit=1)
-        if open_payment:
+        payment_greater_than_reconcile = line_obj.search(
+            payment_search + [('credit', '>', reconcile_amount)],
+            order='date asc', limit=1)
+        if payment_greater_than_reconcile:
             # Split the payment move line to isolate reconcile amount
-            (open_payment | move_lines).split_payment_and_reconcile()
+            (payment_greater_than_reconcile | move_lines) \
+                .split_payment_and_reconcile()
+            return True
         else:
             # Group several payments to match the invoiced amount
             # Limit to 12 move_lines to avoid too many computations
-            open_payments = line_obj.search(payment_search[:-1], limit=12)
+            open_payments = line_obj.search(payment_search, limit=12)
             # Search for a combination giving the invoiced amount
             # https://stackoverflow.com/questions/34517540/
             # find-all-combinations-of-a-list-of-numbers-with-a-given-sum
-            credits = open_payments.mapped('credit')
             matching_lines = line_obj
-            must_split_move_line = False
-            for i in range(2, len(credits)):
-                for combination in itertools.combinations(credits, i):
-                    if sum(combination) == total_amount:
-                        prev_index = 0
-                        for amount in combination:
-                            index = credits[prev_index:].index(amount)
-                            matching_lines += open_payments[prev_index:][index]
-                            prev_index += index + 1
-                        break
-                if matching_lines:
-                    break
+            all_combinations = \
+                (combination for n in range(2, len(open_payments) + 1)
+                 for combination in itertools.combinations(open_payments, n))
+            for combination in all_combinations:
+                total = sum(payment.credit for payment in combination)
+                if total == reconcile_amount:
+                    for payment in combination:
```

Comment:
Why not directly do `(combination | move_lines).reconcile()`

Reply:
I guessed it was there for a reason. Don't both operands have to be RecordSets?

Reply:
It was my same mistaking, I didn't see combination is a tuple and not a recordset. So I think it's good now.